File: TeamGameFolder/team_game.py
#  IB GLG, 7th period, Adventure RPG Game - Improved Version

# GLG = 🎀
# IB = 😊

# Step by step  plan:
# Defines each of the needed variables, including room names
# Plans the rooms
# Makes you able to decide what/where to go.
# Loops that choice
# Boss
# Win or lose
# Combines and plays all functions within a function (probably called 'main').
# Calls the main function

import random

#------------------
# CONSTANTS | Written by 🎀 | Edited by 😊
#------------------

# Defines some things that will never change throughout the game.

MAX_INVENTORY = 5

weapons = {
"fists": 1,
"stick": 2,
"sword": 4,
"magic wand": 6
}

rooms = ["forest", "cave", "river", "temple"]

# -----------------------------------
# RESET GAME | Written by 😊 | Edited by
# -----------------------------------

# This is how we reset the game if you die.

def reset_game():

    global health, gold,  inventory, current_room, game_running

    health = 12
    gold = 0
    inventory = []
    current_room = "forest"
    game_running = True

# -----------------------------------
# INSTRUCTIONS | Written by 😊 | Edited by
# -----------------------------------

# Opening title page! Shows you the commands for the game.

def show_instructions():
    print("\n===================================")
    print("     ADVENTURE RPG GAME")
    print("===================================")
    print("Commands:")
    print(" north / south / east / west")
    # print(" look")
    print(" inventory")
    print(" map")
    print(" drop")
    print(" shop")
    print(" quit")
    print("===================================")

# -----------------------------------
# MAP prints a simple map showing where you are. | Written by 😊 | Edited by
# -----------------------------------

def show_map():

    pos = current_room # pos is position


    def mark(room):
        if room == pos:
            return "x"
        return " "
    
    print("\nMAP:\n")

    print(f" [C{mark('cave')}]---[T{mark('temple')}]")
    print("   |")
    print(f" [F{mark('forest')}]---[💰R{mark('river')}]")

    print("\nLegend: \nX = You \nC = cave \nT = Temple \nF = forest \nR = river \n💰 = potion shop")


# -----------------------------------
# BEST WEAPON | Written by 😊 | Edited by
# -----------------------------------

# FInds the most powerful weapon in your inventory and automatically uses it to fight

def get_best_weapon():

    best = "fists"
    best_power = weapons["fists"]

    for item in inventory:

        if item in weapons:

            if weapons[item] > best_power:
                best = item
                best_power = weapons[item]

    return best

# -----------------------------------
# INVENTORY | Written by 😊 | Edited by
# -----------------------------------

def show_inventory():

    print("\n--------------------")
    print("Health:", health)
    print("Gold: 🪙",gold)
    print("Best Weapon:", get_best_weapon())
    print("Inventory:", inventory)
    print("Items:", len(inventory), "/", MAX_INVENTORY)

# -----------------------------------
# DROP ITEM | Written by 😊 | Edited by
# -----------------------------------

# Lets you drop unwanted items and empty as mush of your inventory as you want.

def drop_item():

    if not inventory:
        print("Inventory is empty.")
        return
    
    print("\nDrop which item?")

    for i in range(len(inventory)):
        print(i+1, "-", inventory[i])

    try:
        choice = int(input(">")) - 1

        if 0<= choice < len(inventory):

            removed = inventory.pop(choice)
            print("Dropped:", removed)

        else:
            print("Invalid choice.")

    except:
        print("Invalid input.")

# -----------------------------------
# SHOP | Written by 🎀 | Edited by 😊
# ----------------------------------

# The potion shop by the river

def potion_shop():
    
    global gold
    
    print("\n---POTION SHOP ---")
    print("potion cost 5 gold.")
    print("your gold:", gold)

    
    if len(inventory) >= MAX_INVENTORY:
        print("Inventory full. Drop something first.")
        return
    
    if gold < 5:
        print("You dummy, your poor, HA HA HA. Go fight monsters to get more Gold")
        return 
    
    buy = input("Buy potion? (yes/no): ").lower().strip()

    if buy == "yes":
        gold -= 5
        inventory.append("potion")
        print("You bought a potion!")
        
    
# -----------------------------------
# CHEST | Written by 😊 | Edited by 🎀
# -----------------------------------

# This shows what the possible contents of the chest are, some of which are influenced by where you are.

def open_chest(room):

    print("\nYou found a chest!")
    
    if room == "forest":
        weapon = "stick"
    
    elif room == "cave":
        weapon = "magic wand"

    elif room == "river":
        weapon = "sword"
    
    else:
        weapon = "sword"
        
    print("Inside is a ", weapon + "!")
    
    if len(inventory) >= MAX_INVENTORY:
        
        print("Inventory full.")

        drop = input("🗑️ Drop item? (yes/no): ").lower().strip()
        
        if drop != "yes":
            return
        
        drop_item()
        
    inventory.append(weapon)

    print("You stored the", weapon + ".")

# -----------------------------------
# COMBAT  | Written by 🎀 | Edited by 😊
# -----------------------------------

# Here is where it has the monsters health and attacks, as well as how you fight ect. This section also rewards you with gold [or winnning the game] after defeating enemies.

def fight_monster(boss=False):
    
    global health, gold


    if boss:
        monster_health = random.randint(30, 36)
        print("\nTHE DARK BOSS APPEARS!")
    else: 
        
        monster_health = random.randint(5, 8)
        print("\nA monster attacks! OH NO!")

    while monster_health > 0 and health > 0:


        print("\nyour Health:", health)
        print("Enemy Health:", monster_health)
        print ("options: fight / run / heal")

        choice = input("> ").lower()

        if choice == "fight":

            weapon = get_best_weapon()
            power = weapons[weapon]

            player_damage =  random.randint(1, power + 2)
            monster_damage = random.randint(1, 3)

            monster_health -= player_damage 
            health -= monster_damage 

            print("you used:", weapon)
            print("you dealt:" , player_damage)
            print ("enemy dealt:", monster_damage)
      
        elif choice == "run":

            if boss: 
                print("You  cannot run from the boss!")

            else: 

                if random.randint(1,2) == 1:
                    print("You escaped!")
                    return
                else:
                    damage = random.randint(1, 2)
                    health -= damage 
                    print("Failed escape! Lost", damage)


        elif choice == "heal":

            if "potion" in inventory:
                
                inventory.remove("potion")
                health += 4

                print("Healed! health:", health)

            else:
                print("no potions!")

        else:
            print("Invalid command.")

    if monster_health <= 0:
    
      if boss:
        boss_reward = random.randint(2000, 99999)
        gold += boss_reward
        print("\n💰🪙💰🪙💰🪙💰🪙YOU WIN 💰 ", boss_reward, "🪙 GOLD!💰🪙💰🪙💰🪙💰🪙")
        print ("\nYOU BEAT THE GAME! 🏆")
        return "win"

      else:
        reward =  random.randint(3, 6)
        gold += reward 
        print("Monster defeated!") 
        print("Gained", reward, "gold.")
        
# -----------------------------------
# RANDOM EVENT  | Written by 🎀 | Edited by 😊
# -----------------------------------

# Here is where it decises what happens each time that you walk into a room. Whether it is being attacked, getting a chest/item, or nothing...

def random_event():
    
    event = random.randint(1, 4)

    if event == 1:

        if random.choice(["gold", "potion"]) == "gold":

            global gold 
            amount = random.randint(2,5)
            gold += amount

            print("you found", amount, "gold!")

        else:

            if len(inventory) < MAX_INVENTORY:

                inventory.append("potion")
                print("you found a potion!")

            else:

                print("Found a potion, but inventory is full.")

    elif event == 2:

        fight_monster()

    elif event == 3:

        print("Nothing happens...")

    elif event == 4:

        open_chest(current_room)
    



# -----------------------------------
# ROOMS  | Written by 🎀 | Edited by 😊
# -----------------------------------

# Defines the rooms, and the options to go from each of the rooms, as well as the 'intro' print line to each room.

def forest():
    print("\nyou are in a scary forest. 👻")
    print("paths: north, east")

def cave():
    print("\nyou are in a dark cave.")
    print("paths: south, east")

def river():
    print ("\nyou stand beside a beautiful river.")
    print("paths: west, north")
    print("A potion stand is here.")

def temple():
    print("\ntyou enter the ancient temple.")

    result = fight_monster(boss=True)

    if result == "win":
        return True 

    return False 





# -----------------------------------
# MAIN  | Written by 😊 | Edited by 🎀 
# -----------------------------------

# Puts all of the previous functions together and actually runs it.

def main():

    global current_room, game_running

    reset_game()
    show_instructions()

    while game_running:

        if health <=0:

            print("\nGAME OVER 💀")

            again = input("Play again? ").lower().strip()

            if again == "yes":
                reset_game()
                continue
            else:
                break

        if current_room == "forest":
            forest()

        elif current_room == "cave":
            cave()

        elif current_room == "river":
            river()

        elif current_room == "temple":

            win = temple()

            if win:

                again = input("Play again? ").lower().strip()

                if again == "yes":
                    reset_game()
                    continue
                else:
                    break

        command = input("\nWhat do you do? ").lower().strip()

        moved = False

        # BASIC (Basic commands such as looking at inventory, map, and quitting the game.)

        if command == "quit":
            break

        elif command == "inventory":
            show_inventory()

        elif command == "map":
            show_map()

        elif command == "drop":
            drop_item()

        elif command == "shop":

            if current_room == "river":
                potion_shop()
            else:
                print("No shop here.")

        # MOVEMENT (Allows and restricts your character movement between rooms/places [only one restriction, needing the magic wand to enter the temple to fight the boss])

        elif command == "north":

            if current_room == "forest":
                current_room = "cave"
                moved = True

            elif current_room == "river":

                if "magic wand" in inventory:
                    current_room = "temple"
                    moved = True
                else:
                    print("Barrier! Need magic wand. (P.S. Try the caves)")

            else:
                print("Can't go that way.")


        elif command == "south":

            if current_room == "cave":
                current_room = "forest"
                moved = True
            else:
                print("Can't go that way.")


        elif command == "east":

            if current_room == "forest":
                current_room = "river"
                moved = True
 
            elif current_room == "cave":

                if "magic wand" in inventory:
                    current_room = "temple"
                    moved = True
                else:
                    print("Barrier! Need magic wand. (P.S. Try the caves)")

            else:
                print("Can't go that way.")
        
        elif command == "west":
            if current_room == "river":
                current_room = "forest"
                moved = True
            else:
                print("Can't go that way.")

        else:
            print("Unknown command.")

        if moved and current_room != "temple":
            random_event()


# -----------------------------------
# START  | Written by 😊
# -----------------------------------

main()

File: C/TeamGameFolderC/team_game.c
// add function to make a monster in the forest be the black knight
//show an 'X' at where you are on the map


// =====================================================
// THE HOLY GRAIL
// Written by: IB and RH
// 7th Period
// =====================================================

#include <stdio.h>
#include <stdlib.h>
#include <string.h> // Added for string handling functions like strcmp, strcpy, and strcspn (for input handling and inventory management)
#include <time.h> // Added for time function, which is used to seed the random number generator (for more varied gameplay, like damage and item drops)
#include <ctype.h> // Added for tolower function, which converts characters to lowercase (used in input handling)
#include <stdbool.h> // Added for bool, rather than having to use int (0 or 1) for true/false

// ---------------------------
// Global Constants | Written by RH | Edited by IB
// ---------------------------
#define MAX_INVENTORY 5
#define MAX_HEALTH 12

// ---------------------------
// Global Variables | Written by RH | Edited by IB
// ---------------------------
int health;
int potions;
int inventory_count;
int weapon_power;

char current_weapon[20];

bool has_wand;
bool forest_hint_seen;
bool boss_defeated;


// ---------------------------
// Helper Functions | Written by IB | Edited by 
// ---------------------------

// Written by IB | Edited by
// Remove the newline character left by fgets
void remove_newline(char str[]) {
    str[strcspn(str, "\n")] = '\0'; // strcspn is from string.h and returns the index of the first occurrence of the second string in the first string (or the length of the first string if it is not found). This effectively replaces the newline character with a null terminator, removing it from the string. (In other words, it finds the newline character(\n) and replaces it with a null terminator, which ends the string there.) https://www.w3schools.com/c/ref_string_strcspn.php
}

// Written by IB | Edited by
// Convert a string to lowercase
void to_lowercase(char str[]) {
    for (int i = 0; str[i] != '\0'; i++) {
        str[i] = tolower((unsigned char)str[i]); // tolower is from ctype.h (which is why I added that as well) and converts a character to lowercase. https://www.geeksforgeeks.org/c/tolower-function-in-c/
    }
}

// Written by IB | Edited by
// Add an item to the inventory if there is space
bool add_item(const char item[]) {
    if (inventory_count >= MAX_INVENTORY) {
        printf("Your inventory is full! You leave the %s behind.\n", item);
        return false;
    }

    inventory_count++;
    printf("You picked up: %s\n", item);

    // Potions increase potion count
    if (strcmp(item, "potion") == 0) { // strcmp compares two strings and returns 0 if they are the same. (like you have in your notes...) https://www.w3schools.com/c/ref_string_strcmp.php
        potions++;
    }
    // Stick upgrades weapon if stronger
    else if (strcmp(item, "stick") == 0) {
        if (weapon_power < 2) {
            weapon_power = 2;
            strcpy(current_weapon, "stick"); // strcpy is from string.h and copies the second string into the first string (which is why we needed string.h). https://www.w3schools.com/c/ref_string_strcpy.php
            printf("You equip the stick.\n");
        }
    }
    // Sword upgrades weapon if stronger
    else if (strcmp(item, "sword") == 0) {
        if (weapon_power < 4) {
            weapon_power = 4;
            strcpy(current_weapon, "sword");
            printf("You equip the sword.\n");
        }
    }
    // Magic wand upgrades weapon if stronger and unlocks temple
    else if (strcmp(item, "magic wand") == 0) {
        has_wand = true;
        if (weapon_power < 6) {
            weapon_power = 6;
            strcpy(current_weapon, "magic wand");
            printf("You equip the magic wand.\n");
        }
    }

    return true;
}

// Written by IB | Edited by
// Use one potion to heal the player
bool use_potion() {
    if (potions <= 0) {
        printf("You don't have any potions!\n");
        return false;
    }

    if (health >= MAX_HEALTH) {
        printf("You are already at full health.\n");
        return false;
    }

    potions--;
    inventory_count--;

    health += 4;
    if (health > MAX_HEALTH) {
        health = MAX_HEALTH;
    }

    printf("You used a potion and restored health!\n");
    printf("Health: %d/%d | Potions left: %d\n", health, MAX_HEALTH, potions);

    return true;
}

// ---------------------------
// Game Setup / Display | Written by IB | Edited by 
// ---------------------------

// Written by IB | Edited by
// Reset all game variables at the start of a new game
void reset_game() {
    health = MAX_HEALTH;
    potions = 0;
    inventory_count = 0;
    weapon_power = 1;

    strcpy(current_weapon, "fists");

    has_wand = false;
    forest_hint_seen = false;
    boss_defeated = false;
}

// Written by IB | Edited by
// Show instructions for the player
void show_instructions() {
    printf("\n===== HOW TO PLAY =====\n");
    printf("Explore the forest, cave, river, and temple.\n");
    printf("Find better weapons and potions to survive.\n");
    printf("You may need to revisit places to discover new items.\n");
    printf("Commands: north, south, east, west, inventory, heal, quit\n");
    printf("In battle: fight, heal, run\n");
    printf("Goal: Defeat the final boss in the temple!\n");
    printf("========================\n\n");
}
// Written by IB | Edited by
// Show current player stats
void show_stats() {
    printf("\n----- PLAYER STATS -----\n");
    printf("Health: %d/%d\n", health, MAX_HEALTH);
    printf("Weapon: %s (Power %d)\n", current_weapon, weapon_power);
    printf("Potions: %d\n", potions);
    printf("Inventory Slots Used: %d/%d\n", inventory_count, MAX_INVENTORY);
    printf("------------------------\n");
}

/*
//Written by IB | Edited by
// Updates the map as to show you where you are

pos = current_room

char mark(room) {
    if (room ==)
}
*/


// Written by IB | Edited by
// Show the simple map
void show_map() {
    printf("\n===== MAP =====\n");
    printf("        [North: Temple]\n");
    printf("[West: Forest] [Start] [East: Cave]\n");
    printf("        [South: River]\n");
    printf("================\n");
}


// ---------------------------
// Battle System | Written by IB | Edited by RH
// ---------------------------
bool fight_monster(bool boss) {
    int enemy_health;
    int enemy_damage_min;
    int enemy_damage_max;

    if (boss) {
        enemy_health = rand() % 5 + 24; // 24 through 28
        enemy_damage_min = 2;
        enemy_damage_max = 4;

        printf("\n*** FINAL BOSS APPEARS! ***\n");
        printf("This enemy looks powerful... you hope you came prepared.\n");
    } else {
        enemy_health = rand() % 4 + 5; // 5 to 8
        enemy_damage_min = 1;
        enemy_damage_max = 3;

        printf("\nA monster attacks!\n");
    }

    while (enemy_health > 0 && health > 0) {
        char action[20];

        printf("\nEnemy Health: %d\n", enemy_health);
        printf("Your Health: %d/%d\n", health, MAX_HEALTH);
        printf("Choose action (fight/heal/run): ");
        
        fgets(action, sizeof(action), stdin);
        remove_newline(action);
        to_lowercase(action);
        
        if (strcmp(action, "fight") == 0) {
            int player_damage = rand() % (weapon_power + 2) + 1;
            enemy_health -= player_damage;
            printf("You hit for %d damage!\n", player_damage);
            
            if (enemy_health > 0) {
                int enemy_damage = rand() % (enemy_damage_max - enemy_damage_min + 1) + enemy_damage_min;
                health -= enemy_damage;
                printf("The enemy hits you for %d damage!\n", enemy_damage);
            }
        } else if (strcmp(action, "heal") == 0) {
            if (use_potion()) {
                int enemy_damage = rand() % (enemy_damage_max - enemy_damage_min + 1) + enemy_damage_min;
                health -= enemy_damage;
                printf("The enemy attacks while you heal for %d damage!\n", enemy_damage);
            }
        } else if (strcmp(action, "run") == 0) {
            if (boss) {
                printf("You cannot run from the final boss! As you turn away, he devours you whole!\n");
                health = 0;
            } else {
                printf("You escaped!\n");
                return true;
            }
        } else {
            printf("Invalid action\n");
        }
    }
    
    if (health <= 0) {
        printf("\nYou were defeated...\n");
        return false;
    }
    
    if (boss) {
        printf("\n*** You defeated the final boss and won the game! ***\n");
        boss_defeated = true;
    } else {
        printf("\nYou defeated the monster!\n");
    }

    return true;
}

// ---------------------------
// Room Functions | Written by RH | Edited by 
// ---------------------------
// forest room
bool forest_room() {
    printf("\nYou enter the Forest.\n");

    if (!forest_hint_seen){
        printf("\nA paper pinned to a tree reads: \"Old places may hide new treasures.💰\"\n");
        forest_hint_seen = true;
    }
    
    int event = rand() % 3;
    
    if (event == 0) {
        printf("\nYou find a potion near a tree stump.\n");
        add_item("potion");
    }
    else if (event == 1){
        printf("\nYou encounter the black knight. \n");
        if (!fight_monster(false)) {
            return false;
        }
    }
    else {
        printf("\nThe forest is quiet... too quiet.🦗🦗\n");
    }
    return true;
}

// cave room
bool cave_room() {
    printf("\nYou enter the cave,👻 it looks spooky.👻👻\n"); // I added it looks spooky for affects
        
    int event = rand() % 4;
       
    if (event == 0) {
         printf("\nYou find a sturdy stick on the ground,🥀 It's not the best.\n");
          add_item("stick");
    } else if (event ==1) {
        printf("\nYou discover an old sword stuck in rocks.🗡️\n"); // I changed the frasing a bit
        add_item("sword");
    } else if (event ==2) {
        printf(" \nDeep in the shadows you find... A MAGIC WAND!!🪄🪄\n"); // changed words to sound more fun
        add_item("magic wand");
    } else {
        printf("\nThe cave chest is empty.\n");
        printf("A note inside reads: \"Raided... but treasures return to those who check again.\"\n");
    }
    return true;
}

// River room
bool river_room() {
    printf("\nYou walk to the river.\n");
        
    int event = rand() % 4;
        
    if (event == 0) {
        printf("\nYou find a potion floating near the shore.\n");
        add_item("potion"); 
    }
    else if (event == 1) {
        printf("\nSomething lurks near the water... It's a gnome! He runs towards you.\n");
        if (!fight_monster(false)) {
            return false;
        }
    }
    else if (event == 2) {
        printf("\nSomething lurks near the water... It's a gnome! He runs away.\n");
    }
    else {
        printf("\nThe river is calm. You take a short rest.\n");
    }
        
    return true;
}
// Temple room
bool temple_room() {
    printf("\nYou arrive at the ancient temple of the gods.\n");

    if (!has_wand) {
        printf("\nA magical barrier blocks you from entering the temple.\n");
        printf("\nA glowing poneglyph says: \"Only the wand lets past.\"\n");
        return true;
    }
    printf("\nYou cast a spell and the barrier disappears before your eyes.\n");
    printf("\nYou step into the temple\n");

    return fight_monster(true);

}


// ---------------------------
// Main Game Loop | Written by  | Edited by 
// ---------------------------

// Written by IB | Edited by
// Run one full game
void play_game() {
    char command[20];
    bool game_running = true;

    reset_game();
    show_instructions();

    while (game_running && health > 0 && !boss_defeated) {
        show_map();
        show_stats();

        printf("\nWhere do you want to go? ");
        fgets(command, sizeof(command), stdin);
        remove_newline(command);
        to_lowercase(command);

        if (strcmp(command, "north") == 0) {
            // If player has the wand, they may enter the temple and fight the boss
            if (has_wand) {
                if (!temple_room()) {
                    game_running = false;
                }
            } else {
                // Otherwise, just show the barrier message
                temple_room();
            }
        }
        else if (strcmp(command, "south") == 0) {
            if (!river_room()) {
                game_running = false;
            }
        }
        else if (strcmp(command, "east") == 0) {
            if (!cave_room()) {
                game_running = false;
            }
        }
        else if (strcmp(command, "west") == 0) {
            if (!forest_room()) {
                game_running = false;
            }
        }
        else if (strcmp(command, "inventory") == 0) {
            printf("\nYou are carrying:\n");
            printf("- Weapon: %s\n", current_weapon);
            printf("- Potions: %d\n", potions);
            printf("- Inventory slots used: %d/%d\n", inventory_count, MAX_INVENTORY);
        }
        else if (strcmp(command, "heal") == 0) {
            use_potion();
        }
        else if (strcmp(command, "quit") == 0) {
            printf("You leave the adventure behind.\n");
            game_running = false;
        }
        else {
            printf("Invalid command.\n");
        }
    }

    if (boss_defeated) {
        printf("\nCongratulations, adventurer!\n");
    }
    else if (health <= 0) {
        printf("\nGame Over.\n");
    }
}

// ---------------------------
// Main Function | Written by  | Edited by 
// ---------------------------

// Written by IB | Edited by
int main() {
    char play_again[10];

    srand(time(NULL));

    printf("=====================================\n");
    printf("           THE HOLY GRAIL\n");
    printf("=====================================\n");

    do  {
        play_game();
        printf("\nWould you like to play again? (yes/no): ");
        fgets(play_again, sizeof(play_again), stdin);
        remove_newline(play_again);
        to_lowercase(play_again);
    } while (strcmp(play_again, "yes") == 0);
    
    printf("\nThanks for playing!\n");
    return 0;
}